p2-maven-plugin/src/main/java/org/eclipse/tycho/p2maven/tools/P2RepositoryDataManipulator.java

```diff
@@ -0,0 +1,193 @@
+					this::mirrorMetadataRepository, r -> r.executeBatch(m -> modification.accept(r), null));
+		}
+		xzCompress(repository, outputLocation);
+	}
+
+	private <T, R extends IRepository<T>> void modifyOutputRepository(R repository, Path outputLocation,
+			BiFunction<R, Path, R> copyAction, Consumer<R> modification) throws ProvisionException {
+		if (outputLocation != null) {
+			R localRepository = copyAction.apply(repository, outputLocation);
+			modification.accept(localRepository);
+		} else if (repository.isModifiable()) {
+			modification.accept(repository);
+		} else {
+			throw new IllegalStateException(
+					"Repository is not modifable and no output location is defined: " + repository.getLocation());
+		}
+	}
+
+	/**
+	 * Creates a local mirror only of the <b>data</b> of the given
+	 * {@link IMetadataRepository}, i.e. its {@code artifacts.xml/jar/xml.xz} files.
+	 */
+	private IArtifactRepository mirrorArtifactRepositoryData(IArtifactRepository repository, Path targetLocation) {
+		try {
+			IArtifactRepository mirror = manager.createLocalArtifactRepository(targetLocation, repository.getName(),
+					repository.getProperties());
+			IStatus result = mirror.executeBatch(m -> {
+				// Copy artifacts to modifiable output repository
+				repository.query(ArtifactKeyQuery.ALL_KEYS, null).stream()
+						.map(a -> repository.getArtifactDescriptors(a)).forEach(a -> mirror.addDescriptors(a, null));
+				// Copy (potentially customized) mapping rules
+				if (repository instanceof SimpleArtifactRepository simpleSource
+						&& mirror instanceof SimpleArtifactRepository simpleTarget) {
+					simpleTarget.setRules(simpleSource.getRules()); // Copy artifact mapping rules
+				}
+			}, null);
+			assertNoError(result, targetLocation);
+			// TODO: remove repository from manager again?
+			return mirror;
+		} catch (ProvisionException e) {
+			throw new IllegalStateException(e);
+		}
+	}
+
+	/** Creates a local mirror of the given {@link IMetadataRepository}. */
+	private IMetadataRepository mirrorMetadataRepository(IMetadataRepository repository, Path targetLocation) {
+		try {
+			IMetadataRepository mirror = manager.createLocalMetadataRepository(targetLocation, repository.getName(),
+					repository.getProperties());
+			IStatus result = mirror.executeBatch(m -> {
+				// Copy units to modifiable output repository
+				repository.query(QueryUtil.ALL_UNITS, null).stream().map(List::of).forEach(mirror::addInstallableUnits);
+			}, null);
+			assertNoError(result, targetLocation);
+			// TODO: remove repository from manager again?
+			return mirror;
+		} catch (ProvisionException e) {
+			throw new IllegalStateException(e);
+		}
+	}
+
+	private static void assertNoError(IStatus status, Path location) {
+		if (!status.isOK() && status.matches(IStatus.ERROR)) {
+			throw new IllegalStateException("Failed to create output repository at: " + location + "\n" + status,
+					status.getException());
+		}
+	}
```

Comment:
This could be moved to the P2RepositoryManager as well, as this might be relevant for other tasks too.

Reply:
Unfortunately this cannot be used here as it resides in `tycho-core` which depends on `p2-maven-plugin`. So it would require a cycle dependency.

Reply:
we can move it probably to `tycho-spi` or `tycho-api` module.

Reply:
Moved it to `tycho-spi `, but still had to add a method that asserts the severity of a status is below a certain level.
But this is now reusable.

Reply:
> Moved it to tycho-spi , but still had to add a method that asserts the severity of a status is below a certain level.

Can we have this as a separate PR to keep this one easier to review? I think that change can be merged right away (if not tests fail).

Reply:
Done with:
- https://github.com/eclipse-tycho/tycho/pull/5225

Auto-merge is enabled.